Port test_initial_ap_config to boardfarm for prplWRT device

abelog opened the pull request:
## Description
This PR partially resolves #1072 .
Main goal is to bring InitialApConfig to RAX40.

**Prerequisite**:
1.  Interface, which has RAX40 connected, is included in docker bridge.
1.  Name of this docker network is stated in `boardfarm_plugins/boardfarm_prplmesh/prplmesh_config.json` 

This network is used for dockerized controller -> RAX40's agent. In order to add iface to the bridge, `sudo` powers are required. Currently, we do not plan to run `bft` with `sudo` power. Thus, docker network preparation and manual include of interface is required.

## Changes
- `prplWRT` device class to handle RAX40
- prplMesh specific entities for `environment.py`
- `rax40-1`RAX40 board added to boardfarm configuration

abelog committed:
tests: add device to boardfarm config

Add "prplWRT" device "rax40-1" to boardfarm config.

In order to this config to work, you have to:
1. Create docker bridge network with name "prplMesh-net-rax40-1".
`docker network create --driver=bridge prplMesh-net-rax40-1`
2. Include interface connected to the board in docker network.
`ip link set $RAX40_IFACE master $DOCKER_BRIDGE`
3. Config relies on RAX40 consoles to be at "/dev/ttyUSB0". You may have
to adjust this path to fit your system.

Those steps are required, as regular user is not permitted to attach
interfaces to the bridge.

Signed-off-by: Anton Bilohai <a.bilohai@inango-systems.com>

abelog committed:
tests: boardfarm: add prplWRT device

Add prplWRT class which represents device running prplWRT with prplMesh
installed. This class has required methods to init the board and execute
InitialApConfig test on board.

Signed-off-by: Anton Bilohai <a.bilohai@inango-systems.com>

abelog committed:
tests:  environment.py: complete prplWRT entities

prplWRT entities are not debugged, as there was no way to test them at
the moment of commit.

`_device_wait_for_log` method: count in start_line given to it, format
return as it's expected by callers.
`ALEntityPrplWrt` constructor: get main bridge name from config. Parse
output of console in Python to be less dependent on board.
`RadioHostapd` constructor: remove usage of `ip -o` as `-o` option is
bugged on RAX40 and mutes output of `ip` command.

Signed-off-by: Anton Bilohai <a.bilohai@inango-systems.com>

abelog committed:
tests: boardfarm: InitialApConfig - too greedy regex search

Currently, test is trying to match r".* Controller configuration \(WSC
M2 Encrypted Settings\)") which is greedy and captures all chars before
key message. Chars captured by `.*` are not important for us.

Get rid of `.*` in the beginning of regex. Reorder expected logs to
match them in right order.

Signed-off-by: Anton Bilohai <a.bilohai@inango-systems.com>

MuteSpirit commented on tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_prplwrt.py:
```diff
@@ -6,19 +6,27 @@
+import sys
+
 from .prplmesh_base import PrplMeshBase
+from boardfarm.devices import openwrt_router, connection_decider
```
To use class name without module name, for example, `openwrt_router.OpenWrtRouter` try
`from boardfarm.devices.openwrt_router import OpenWrtRouter`

abelog replied:
Fixed!

MuteSpirit commented on tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_prplwrt.py:
```diff
@@ -67,3 +103,37 @@ def touch(self):
+        self._prplMesh_exec("certification_mode agent")
+        self.sendline("/etc/init.d/prplmesh certification_mode agent")
+        self.expect_exact("CAC timer expired", timeout=120)
+        self.expect_exact("device br-lan entered promiscuous mode", timeout=40)
```
Name of bridge, `br-lan` in this case, is depend on image settings
I think that it'll be better to take it from setup configuration

abelog replied:
Both 2 logs are actually specific to RAX40 and driver it uses. Till those logs, device isn't yet ready.
Only good solution i see, is to move that code to separate device class which inherits `prplMesh`.

MuteSpirit commented on tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_prplwrt.py:
```diff
@@ -67,3 +103,37 @@ def touch(self):
+    def prplMesh_start_agent(self) -> bool:
+        """Start prplMesh in certification_mode agent. Return true if done."""
+        self._prplMesh_exec("certification_mode agent")
+        self.sendline("/etc/init.d/prplmesh certification_mode agent")
```
Both lines perfome the same shell command on board
Why is it needed?

abelog replied:
Left after debug. Should be removed.

abelog replied:
Fixed.

MuteSpirit commented on tests/boardfarm_plugins/boardfarm_prplmesh/prplmesh_config.json:
```diff
@@ -16,5 +16,22 @@
+        "role": "agent",
+        "docker_network": "prplMesh-net-rax40-1",
+        "connection_type": "local_serial",
+        "conn_cmd": "cu -s 115200 -l /dev/ttyUSB0",
```
Assume that two UARTs are connected to the same PC
* 1st one device is `/dev/ttyUSB0`
* 2nd one device is `/dev/ttyUSB1`
But it's not guaranteed that after PC reboot they will not change their devices names and become:
* 1st one device is /dev/ttyUSB**1**
* 2nd one device is /dev/ttyUSB**0**

But if you will use path to serial ports like `/dev/serial/....` and will not replug UARTs into another USB ports then paths to the serial devices will not be changed

abelog replied:
This is a generic config. AFAIK, path in `/dev/serial/by-path/...` can be very different between workstations. That why `/dev/ttyUSB0` was selected.

rmelotte replied:
What we did in the testbed we use in CI is to use /dev/<device_name>, with a udev rule to create the correct symlink.
Maybe we could also do the same here?

abelog replied:
We can. `dev/ttyUSB0` is used because it's generic. What is serial path of RAX40 on CI machine?

arnout replied:
Serial path of device netgear-rax40-1 is /dev/netgear-rax40-1. You can take a guess as to the path of device turris-omnia-1 :-)

MuteSpirit commented on tests/boardfarm_plugins/boardfarm_prplmesh/tests/initial_ap_config.py:
```diff
@@ -22,13 +26,14 @@ def runTest(self):
+        for dev in test.dev:
+            if dev.agent_entity:
+                print("Sniffer - stop")
+                dev.agent_entity.device.send('\003')
```
Maybe use special pexpect function for that `dev.agent_entity.device.sendcontrol('c')` ?

abelog replied:
For unknown to reason, `sendcotrol('c')` isn't stable. As result,  `tail -f` is not stopped.
`device.send('\003')` works always.

MuteSpirit commented on tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_prplwrt.py:
```diff
@@ -67,3 +103,37 @@ def touch(self):
+
+        return IPv4Network(inspect_json[0]["IPAM"]["Config"][0]["Subnet"])
+
+    def set_iface_to_bridge(self, iface: str) -> bool:
```
By my opinion more readable func name is `add_iface_into_bridge`

MuteSpirit commented on tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_prplwrt.py:
```diff
@@ -67,3 +103,37 @@ def touch(self):
+    def set_iface_ip(self, iface: str, ip: IPv4Network) -> bool:
+        """Set interface IPv4 address."""
+        self.sendline("ip a add {}/{} dev {}".format(ip[+245], ip.prefixlen, iface))
+        match = self.expect(self.prompt, timeout=10)
```
This function does not set any IP on interface
Its set IP from passed subnet with last octet `245`
If you want to have general function then it's needed to pass IP and subnet prefix via function arguments and not generate IP inside

abelog replied:
Fixed.
Changed method args to accept IP and subnet prefix separately.

MuteSpirit commented on tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_prplwrt.py:
```diff
@@ -27,27 +35,55 @@ def __init__(self, *args, **kwargs):
         It is used by boardfarm to indicate that spawned device instance is ready for test
         and also after test - to insure that device still operational.
         """
+        return True
```
Please add a comment why did you make this function a stub

abelog replied:
In general, boardfarm uses this method `ping` alike. It executes it in parallel to the test to make sure that device isn't dead. As we are using serial connection, input from status check was mixed up with input from other interacting with console methods.

rmelotte replied:
>As we are using serial connection, input from status check was mixed up with input from other interacting with console methods

This look like a good reason to use the SSH connection instead?

abelog replied:
SSH is now used as primary connection.

MuteSpirit commented on tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_prplwrt.py:
```diff
@@ -27,27 +35,55 @@ def __init__(self, *args, **kwargs):
-                ["OPERATIONAL", self.device.pexpect.EOF, self.device.pexpect.TIMEOUT],
+                ["operational", "FAIL", pexpect.EOF, pexpect.TIMEOUT],
                 timeout=10)
         if match == 1:
```
It's not enough to found `operational` in output because it's possible that controller is operational but one/both agents - no
All of cases - "FAIL", pexpect.EOF, pexpect.TIMEOUT - mean `False` result

MuteSpirit commented on tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_prplwrt.py:
```diff
@@ -67,3 +103,37 @@ def touch(self):
+            subprocess.run(('docker', 'network', 'create', docker_network), check=True,
+                           stdout=subprocess.DEVNULL)
+            # Inspect again, now raise if it fails (check=True).
+            inspect_raw= subprocess.run(docker_network_inspect_cmd, check=True,
```
Missed space before assignment

abelog replied:
Fixed.

MuteSpirit commented on tests/environment.py:
```diff
@@ -189,16 +189,23 @@ def _docker_wait_for_log(container: str, program: str, regex: str, start_line: i
+    device.expect(regex, timeout=timeout_set)
+    match = device.match.group(0)
+    # Send Ctrl-C to interrupt tail -f
+    device.send('\003')
```
Why not `device.sendcontrol('c')`? It's more readable

abelog replied:
For unknown to reason, `sendcotrol('c')` isn't stable. As result,  `tail -f` is not stopped.
`device.send('\003')` works always.

MuteSpirit commented on tests/environment.py:
```diff
@@ -189,16 +189,23 @@ def _docker_wait_for_log(container: str, program: str, regex: str, start_line: i
+        device.sendline("grep -n \"{}\" {}".format(match, log_path))
+        # Typical output of grep -n from log: "line_num:severity"
+        # this regex has to capture just number of line in log
+        device.expect(r"(?P<line_number>[0-9]+):[A-Z]+\s[0-9]", timeout=timeout_set)
```
Will it find last or first match of regex passed in arguments?

MuteSpirit commented on tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_prplwrt.py:
```diff
@@ -27,27 +35,55 @@ def __init__(self, *args, **kwargs):
+        self.consoles = [self]
+        self.logfile_read = sys.stdout
+
+        self.wan_network = self.get_docker_subnet(self.docker_network)
```
Are you expecting that any boards will be connected with test PC only with WAN port and not with LAN port or both LAN and WAN ports?
I think it's not common case

arnout commented on tests/boardfarm_plugins/boardfarm_prplmesh/prplmesh_config.json:
```diff
@@ -16,5 +16,22 @@
             }
         ]
+    },
+    "rax40-1": {
```
It would be easier to call the device "netgear-rax40-1" because that's its name in the test setup.

rmelotte commented on tests/boardfarm_plugins/boardfarm_prplmesh/prplmesh_config.json:
```diff
@@ -16,5 +16,22 @@
+        "name": "agent-rax40",
+        "board_type": "prplWRT",
+        "role": "agent",
+        "docker_network": "prplMesh-net-rax40-1",
```
If the container is started with `--network=host` (which a gitlab-runner can also use), the container will use the same network namespace as the host.
Could it be used when there is only one instance of prplMesh running on the host?
If it works, it would avoid the need to configure the docker network beforehand.

abelog replied:
Boardfarm can handle multiple setups at the same time. Flow is quite simple, let's imagine you have 2 RAX40 set up, with names `nighthawk-rax40-1` and` nighthawk-rax40-2`.
`nighthawk-rax40-1` is already busy with ongoing test and you wish to already start some test with` nighthawk-rax40-2`. 2 options are available:
1. Ask bft (executable name of boardfarm) to start test X explicitly on `nighthawk-rax40-2`
1. Ask bft to start test X on `prplWRT` board type

In the second case, boardfarm will start looking for not busy board described in boardfarm config file and will start the test with the first free board.

As boardfarm can handle few tests at the same time, sharing the same controller entity between multiple agents under test may bring some corner cases (i am not sure if it is possible). Usage of docker bridges brings isolation between setups and excludes the possibility of a cross set up interference issues at all.

arnout replied:
This should be explained in the commit message or in the README.

rmelotte commented on tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_prplwrt.py:
```diff
@@ -6,19 +6,27 @@
 
 
-class PrplMeshPrplWRT(OpenWrtRouter, PrplMeshBase):
+class PrplMeshPrplWRT(openwrt_router.OpenWrtRouter, PrplMeshBase):
```
Why is this change needed?

abelog replied:
Wrong "path" was used. To avoid the change, i have to change import in next way:
````
from boardfarm.devices.openwrt_router import OpenWrtRouter
````

abelog replied:
Fixed.

rmelotte commented on tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_prplwrt.py:
```diff
@@ -6,19 +6,27 @@
     """prplWRT burned device with prplMesh installed."""
 
-    model = "prplWRT"
+    model = ("prplWRT")
```
Why does the model need to be a tuple? I know `PrplMeshDocker` also define it as a tuple, but I couldn't find the information anywhere.

abelog replied:
It's not a must. Either `"prplWRT"` and `("prplWRT")` will work.

rmelotte commented on tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_prplwrt.py:
```diff
@@ -6,19 +6,27 @@
     prompt = ['root\\@OpenWrt:/#', '/#', '@OpenWrt:/#']
     wan_iface = "eth1"
     uboot_eth = "eth0_1"
+    linesep = "\r"
```
This doesn't seem to be used anywhere. Is it?
Also why is it `\r` and not `\n`?

abelog replied:
This variable is used by `pexpect` internally and is not documented.
I have used `\r` following [example](https://github.com/mattsm/boardfarm/blob/master/boardfarm/devices/cougarpark.py) of `cougarpark.py` device class of boardfarm.

rmelotte commented on tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_prplwrt.py:
```diff
@@ -27,29 +35,57 @@ def __init__(self, *args, **kwargs):
         self.sendline("/etc/init.d/prplmesh status")
         match = self.expect(
-                ["OPERATIONAL", self.device.pexpect.EOF, self.device.pexpect.TIMEOUT],
+                ["operational", "FAIL", pexpect.EOF, pexpect.TIMEOUT],
```
I think I saw it mentioned elsewhere: looking for the "operational" string is not enough.
It's possible that the main agent is operational but not the separate radio agents for example.
A more reliable way to check the state would be to check the status code, as the script is supposed to exit with something different than 0 if it was not operational.
With pexpect I think it would require to call close() every time though (which probably isn't really an issue).

abelog replied:
Already fixed! Have changed regex expression to match whole report of `status` command. So, in order to pass the check, device should have main agent and both radios in "OK" state.

rmelotte commented on tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_prplwrt.py:
```diff
@@ -27,29 +35,57 @@ def __init__(self, *args, **kwargs):
+        if match == 0:
             return True
         else:
             return False
```
```suggestion
    return match == 0
```

abelog replied:
Fixed already/

rmelotte commented on tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_prplwrt.py:
```diff
@@ -67,3 +103,37 @@ def touch(self):
+
+    def set_iface_ip(self, iface: str, ip: IPv4Network) -> bool:
+        """Set interface IPv4 address."""
+        self.sendline("ip a add {}/{} dev {}".format(ip[+245], ip.prefixlen, iface))
```
I don't understand what's going on here.
`set_iface_ip` takes an `ip` argument, but of type IPv4Network. Is it supposed to be an ip or a subnet?
It also always adds 245 to the IP it receives. Is the devices forced to have an IP ending with 245?

abelog replied:
Yes. This part was updated, but if no IP/hostname of device is set in boardfarm config - it will be forced to have IP ending with 245.

rmelotte commented on tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_prplwrt.py:
```diff
@@ -27,29 +35,57 @@ def __init__(self, *args, **kwargs):
+
+        self.wan_network = self.get_docker_subnet(self.docker_network)
+        self.set_iface_to_bridge(self.wan_iface)
+        self.set_iface_ip("br-lan", self.wan_network)
```
Why would we always have to set a specific IP?
I guess it's because you want to use an IP in the same subnet as the controller running in a docker container?
This looks unnecessarily restrictive to me. It means we won't be able to run tests with a rax40 unless the controller is running in docker.

If we can run the docker controller with the host network namespace, we can also just use the IP the device already has, or if we needed to, put them in a configuration file.

abelog replied:
> I guess it's because you want to use an IP in the same subnet as the controller running in a docker container?

Exactly.

> This looks unnecessarily restrictive to me. It means we won't be able to run tests with a rax40 unless the controller is running in docker.

Initially, it was planned to have one spawned controller per DUT. Currently, boardfarm is unable to handle non-dockerized controller. We can create a separate boardfarm "device" for it.

rmelotte commented on tests/environment.py:
```diff
@@ -197,16 +197,23 @@ def _docker_wait_for_log(container: str, program: str, regex: str, start_line: i
+        return (True, matched_line, match)
     else:
-        return False
+        return (False, start_line, None)
```
>format return as it's expected by callers.

There is something to be sorted out with _device_wait_for_log.
It returns a free-form tuple that is not mentioned anywhere in the documentation.
(this hasn't been introduced in this PR)

arnout replied:
Well, the idea is in fact to remove it again...
The first (boolean) element is redundant with the third one. The third one is not actually used.
start_line should be replaced with something internal that is stored by checkpoint()

rmelotte commented on tests/environment.py:
```diff
@@ -265,7 +273,8 @@ class RadioDocker(Radio):
         ip_output = agent.command("ip", "-o",  "link", "list", "dev", self.iface_name).decode()
-        mac = re.search(r"link/ether (([0-9a-fA-F]{2}:){5}[0-9a-fA-F]{2})", ip_output).group(1)
+        mac = re.search(r"link/ether (([0-9a-fA-F]{2}:){5}[0-9a-fA-F]{2})",
+                        ip_output).group(1)
```
Are this one and the 2 changes above whitespace changes only? 
If they are they should be done in separate commits, otherwise it forces reviewers to play the [Spot the difference game](https://en.wikipedia.org/wiki/Spot_the_difference).

rmelotte commented on tests/environment.py:
```diff
@@ -365,24 +374,32 @@ class ALEntityPrplWrt(ALEntity):
+        self.bridge_name = re.search(r'bridge_iface=(?P<bridge>.+)\r\n',
+                                     bridge_name_raw).group('bridge')
+
+        # Multiple IPs may be set to same interface. We are interested in the last one, which is set
```
This will also no longer be true if we can use the host network for the Docker container.

rmelotte commented on tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_prplwrt.py:
```diff
@@ -126,14 +126,13 @@ def set_iface_to_bridge(self, iface: str) -> bool:
         self.expect(self.prompt, timeout=10)
 
-    def set_iface_ip(self, iface: str, ip: IPv4Network) -> bool:
+    def set_iface_ip(self, iface: str, ip: IPv4Address, prefixlen :int) -> bool:
```
Ok this fixup makes my previous comment about IPv4Network no longer relevant.
I still think we should use the existing IP if possible.

abelog replied:
Already done! Resolving.

rmelotte reviewed (commented):
Lots of comment, I think some of them should be discussed.
Since Boardfarm is not integrated in CI yet anyway, there is no way to rush this so I'm not approving yet.

abelog committed:
tests: boardfarm: move to ssh as primary connection

Currently, boardfarm tests are using UART (serial) connection as main
channel to interact with the board. Sadly, but UART on RAX40 board
frequently suffers from tty overruns. Hence, tests are frequently failed
due to corrupted input line.

Moving to ssh as primary channel resolves issue above. It's faster, so
chance of meeting input overruns is significantly low. But, via ssh we
are unable to see logs from the Intel WiFi driver. Logs from the driver are
used to detect when device has prplMesh in operational state. As those
logs are not longer shown, we have to come up with other way to check
when prplMesh is operational.

Add fallback mechanism: if IP address of DUTs wan is not set - retrieve
IP subnet from docker network of set up and choose .245 address. This
configuration happens over UART.
Move from UART to SSH as primary communication channel.
Add prplmesh_status_check method to tests, to check if prplMesh on
device is operational.

Signed-off-by: Anton Bilohai <a.bilohai@inango-systems.com>

abelog committed:
tests: boardfarm: docker devices support of InitialAPConfig teardown

Teardown section of InitialAPConfig test introduced by port of this test
to prplWRT devices is incompatible with dockerized devices. It uses
`device.send` method which is not available for docker devices.
Also, prplmesh_status_check method isn't working for docker devices due
to missing link between AlEntityDocker and PrplMeshDocker device.

Following EAFP strategy, surround sent of Ctrl+C in try-catch block.
Add link between AlEntityDocker and PrplMeshDocker as it done for
prplWRT device and restore prplmesh_status_check method functionality.

Signed-off-by: Anton Bilohai <a.bilohai@inango-systems.com>

abelog commented:
Branch was squashed for second cycle of review.

arnout commented on tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_prplwrt.py:
```diff
@@ -27,29 +36,57 @@ def __init__(self, *args, **kwargs):
+                                                        **kwargs)
+        self.connection.connect()
 
+        self.name = "-".join((config.get("name", "rax40"), self.unique_id))
```
I believe "name" is required in the config, so there shouldn't be a need for a default value. If there is one, it should be netgear-rax40, not just rax40.

Why do you append the unique_id? There can be only one boardfarm instance connected to the device, right? So distinguishing them with a unique_id seems pointless.

abelog replied:
> I believe "name" is required in the config, so there shouldn't be a need for a default value. If there is one, it should be netgear-rax40, not just rax40.

Fixed.

> Why do you append the unique_id? There can be only one boardfarm instance connected to the device, right? So distinguishing them with a unique_id seems pointless.

Inheritance from Docker device. Don't see a way we may need it.

arnout commented on tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_prplwrt.py:
```diff
@@ -27,29 +36,57 @@ def __init__(self, *args, **kwargs):
 
+        self.name = "-".join((config.get("name", "rax40"), self.unique_id))
+        self.consoles = [self]
+        self.logfile_read = sys.stdout
```
This could use some explanation (in a comment).

abelog replied:
Added

arnout commented on tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_prplwrt.py:
```diff
@@ -27,29 +36,57 @@ def __init__(self, *args, **kwargs):
+
+        self.wan_network = self.get_docker_subnet(self.docker_network)
+        self.set_iface_to_bridge(self.wan_iface)
+        self.set_iface_ip("br-lan", self.wan_network[+245], self.wan_network.prefixlen)
```
I don't think we should do this kind of thing from boardfarm. The device is assumed to be already configured when boardfarm runs.

arnout commented on tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_prplwrt.py:
```diff
@@ -27,29 +36,57 @@ def __init__(self, *args, **kwargs):
+
+    def prplMesh_check_state(self) -> bool:
+        """ Check prplMesh status. Return True if operational."""
         self.sendline("/etc/init.d/prplmesh status")
```
At some point, this will have to be replaced with a beerocks_cli command on the controller. But not for this PR.

abelog replied:
Created a task: #1443

arnout commented on tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_prplwrt.py:
```diff
@@ -67,3 +104,36 @@ def touch(self):
         """
         pass
+
+    def get_docker_subnet(self, docker_network: str) -> IPv4Network:
```
Since this is a member function, it's not necessary to pass docker_network as an argument.

abelog replied:
Idea was to keep it generalized. But i don't think that there is a case when we need to get the subnet of some unknown docker network.

abelog replied:
Fixed.

arnout commented on tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_prplwrt.py:
```diff
@@ -67,3 +104,36 @@ def touch(self):
+            # Inspect again, now raise if it fails (check=True).
+            inspect_raw = subprocess.run(docker_network_inspect_cmd, check=True,
+                                         stdout=subprocess.PIPE)
+        inspect_json = json.loads(inspect_raw.stdout)
```
This should be refactored with environment._get_bridge_interface (in a separate PR)

arnout commented on tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_prplwrt.py:
```diff
@@ -67,3 +104,36 @@ def touch(self):
+    def set_iface_ip(self, iface: str, ip: IPv4Address, prefixlen :int) -> bool:
+        """Set interface IPv4 address."""
+        self.sendline("ip a add {}/{} dev {}".format(ip, prefixlen, iface))
+        self.expect(self.prompt, timeout=10)
```
As mentioned elsewhere: I don't think these two functions should be needed.

*edit*

After reading the next commit, I now understand this is needed to make sure that the device has an IP address that belongs to the docker network. Still, I think this is taking things too far. Let's just start with the assumption that the device has been set up properly, so if you insist on using a specific network, assume that the interface to the device has already been added to that bridge and that the IP address has been set up properly.

Although even so, I don't see why it would be needed to add the IP address. No IP connectivity is needed between the devices, so why bother?

abelog replied:
> Although even so, I don't see why it would be needed to add the IP address. No IP connectivity is needed between the devices, so why bother?

Do we require IP connectivity between agent and controller, don't we?

arnout replied:
No we don't. MultiAP is purely on Ethernet level.

arnout commented on tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_prplwrt.py:
```diff
@@ -67,3 +104,36 @@ def touch(self):
+        """Start prplMesh in certification_mode agent. Return true if done."""
+        self._prplMesh_exec("certification_mode agent")
+        self.expect_exact("CAC timer expired", timeout=120)
+        self.expect_exact("device br-lan entered promiscuous mode", timeout=40)
```
I get neither of these logs when I execute the command on my rax40...

Why do something new and different here? In the existing tests, we just poll prplmesh_utils.sh status until it's operational or times out, what's wrong with that?

abelog replied:
This is what actually done.  This part of PR is outdated.

arnout commented on tests/environment.py:
```diff
@@ -213,16 +213,23 @@ def _docker_wait_for_log(container: str, program: str, regex: str, start_line: i
-            timeout=timeout)
-    if match == 1:
-        return True
+                         timeout_set: int, start_line: int = 0):
```
Why timeout_set instead of timeout?

The arguments are swapped: start_line comes before timeout.

abelog replied:
Left during debug.

Thanks for noticing second one.

arnout commented on tests/environment.py:
```diff
@@ -213,16 +213,23 @@ def _docker_wait_for_log(container: str, program: str, regex: str, start_line: i
+    """Waits for log matching regex expression to show up."""
+    device.sendline("tail -f -n +{:d} {}".format(start_line, log_path))
+    device.expect(regex, timeout=timeout_set)
+    match = device.match.group(0)
```
This will raise an exception if it timed out, that's not really what we want...

abelog replied:
Created a task: #1444

arnout commented on tests/environment.py:
```diff
@@ -213,16 +213,23 @@ def _docker_wait_for_log(container: str, program: str, regex: str, start_line: i
-        return True
+                         timeout_set: int, start_line: int = 0):
+    """Waits for log matching regex expression to show up."""
+    device.sendline("tail -f -n +{:d} {}".format(start_line, log_path))
```
Tail starts counting at 1, and `tail -n +0` gives the same as `tail -n +1`.
```suggestion
    device.sendline("tail -f -n +{:d} {}".format(start_line + 1, log_path))
```

abelog replied:
Applied!

arnout commented on tests/environment.py:
```diff
@@ -213,16 +213,23 @@ def _docker_wait_for_log(container: str, program: str, regex: str, start_line: i
+        device.sendline("tail -n +{:d} {} | grep -n \"{}\"".format(start_line, log_path, match))
+        # Typical output of grep -n from log: "line_num:severity"
+        # this regex has to capture just number of line in log
+        device.expect(r"(?P<line_number>[0-9]+):[A-Z]+\s[0-9]", timeout=timeout_set)
```
This second one could be done right in the first command, like so:
```
device.sendline("tail -f -n +{:d} {} | awk  '{ print NR ":" $0; }'".format(start_line + 1, log_path))
device.expect(r"^(?P<line_number>[0-9]+):.*" + regex, timeout=timeout)
```
That doesn't give us the match itself, but that's a feature that is currently unused so I think we can simply remove the feature.

arnout commented on tests/environment.py:
```diff
@@ -242,7 +249,8 @@ def __init__(self, name: str, is_controller: bool = False):
             config_file_name = 'beerocks_agent.conf'
         with open(os.path.join(installdir, 'config', config_file_name)) as config_file:
             ucc_port = \
-                re.search(r'ucc_listener_port=(?P<port>[0-9]+)', config_file.read()).group('port')
+                re.search(r'ucc_listener_port=(?P<port>[0-9]+)',
+                          config_file.read()).group('port')
```
This seems to be a spurious change... The line wasn't too long, so why split it?

abelog replied:
To keep formatting same across file.

arnout replied:
OK. In the future, please mention that in the commit message.

arnout commented on tests/environment.py:
```diff
@@ -253,8 +261,8 @@ def __init__(self, name: str, is_controller: bool = False):
         else:
             device_ip_output = self.command(
                 'ip', '-f', 'inet', 'addr', 'show', self.bridge_name)
-            device_ip = re.search(
-                r'inet (?P<ip>[0-9.]+)', device_ip_output.decode('utf-8')).group('ip')
+            device_ip = re.search(r'inet (?P<ip>[0-9.]+)',
+                                  device_ip_output.decode('utf-8')).group('ip')
```
In the future, please do whitespace-only changes in a separate commit. That makes review a lot easier than trying to find out what changed here and then seeing that actually nothing changed.

arnout commented on tests/environment.py:
```diff
@@ -391,66 +400,78 @@ def launch_environment_docker(unique_id: str, skip_init: bool = False, tag: str
-            "grep log_files_path {} | cut -d=\'=\' -f2".format(self.config_file_name))
-
-        ucc_socket = UCCSocket(device_ip, ucc_port)
+        ucc_port_raw = self.command("grep \"ucc_listener_port\" {}".format(self.config_file_name))
+        ucc_port = int(re.search(r'ucc_listener_port=(?P<port>[0-9]+)',
+                                 ucc_port_raw).group('port'))
+        bridge_name_raw = self.command("grep \"bridge_iface\" {}".format(self.config_file_name))
+        self.bridge_name = re.search(r'bridge_iface=(?P<bridge>.+)\r\n',
+                                     bridge_name_raw).group('bridge')
```
(nitpick) it would be better to do a simple `cat` of the config file once, and then do the re.search twice on the same raw output.

arnout commented on tests/environment.py:
```diff
@@ -391,66 +400,78 @@ def launch_environment_docker(unique_id: str, skip_init: bool = False, tag: str
+        RadioHostapd(self, "wlan2")
 
-    def command(self, *command: str) -> bytes:
+    def command(self, command: str) -> bytes:
```
This change is wrong, it has to be `*command`, otherwise prplmesh_command will no longer work.

abelog replied:
Fixed!

arnout commented on tests/environment.py:
```diff
@@ -391,66 +400,78 @@ def launch_environment_docker(unique_id: str, skip_init: bool = False, tag: str
-    def command(self, *command: str) -> bytes:
+    def command(self, command: str) -> bytes:
         """Execute `command` in device and return its output."""
         self.device.sendline(command)
```
So this should be
```suggestion
        self.device.sendline(" ".join(command))
```
Actually the command should be shell-quoted of course.

Also it would be better to use pexpect.replwrap.run_command, if that is available in boardfarm.

abelog replied:
Applied. Thank you.

arnout commented on tests/environment.py:
```diff
@@ -391,66 +400,78 @@ def launch_environment_docker(unique_id: str, skip_init: bool = False, tag: str
     def wait_for_log(self, regex: str, start_line: int, timeout: float) -> bool:
         """Poll the entity's logfile until it contains "regex" or times out."""
         program = "controller" if self.is_controller else "agent"
+        # Multiply timeout by 100, as test sets it in float.
```
Err, both pexpect and the wait_for_log interface have the timeout in seconds, as a float. Where is this centiseconds coming from?

If this is needed, it should be done inside _device_wait_for_log itself.

abelog replied:
AFAIK, pexpect docs [do not state type of timeout](https://pexpect.readthedocs.io/en/stable/api/popen_spawn.html#popenspawn-class). But from default value i guessed that it's int.

> Where is this centiseconds coming from?

They are coming from `prplmesh_base_test.py ` and `initial_ap_config.py` don't define timeout.

arnout replied:
> AFAIK, pexpect docs [do not state type of timeout](https://pexpect.readthedocs.io/en/stable/api/popen_spawn.html#popenspawn-class). But from default value i guessed that it's int.

I tested with a timeout of 1.9, and it waited for about two seconds, so it doesn't round down, so it uses float.

> > Where is this centiseconds coming from?
> 
> They are coming from `prplmesh_base_test.py ` and `initial_ap_config.py` don't define timeout.

Both say that it's in seconds, and also pexpect states that it's in seconds. If you multiply a timeout of .9 with 100, you get a timeout of 90 seconds, which is not what we want...

arnout commented on tests/boardfarm_plugins/boardfarm_prplmesh/tests/initial_ap_config.py:
```diff
@@ -22,13 +26,13 @@ def runTest(self):
+        for dev in test.dev:
+            if dev.agent_entity:
+                print("Sniffer - stop")
+                dev.agent_entity.device.send('\003')
```
This is not explained in the commit message.

It should definitely be a separate commit, with an explanation why.

It does seem like a good idea to stop the sniffer in the teardown function rather than in the test itself. However:
* wired_sniffer should probably not be associated with a device, it should be more "global" (like we have env.wired_sniffer in test_flows). So in this case, it should be associated with the test class itself.
* if it is stopped in teardown, it should also be started from setup.
* the ctrl-C is not explained. I guess you just want to stop whatever is running on the shell, but there should be a comment explaining that.

arnout commented on tests/boardfarm_plugins/boardfarm_prplmesh/tests/prplmesh_base_test.py:
```diff
@@ -16,7 +16,7 @@ class PrplMeshBaseTest(bft_base_test.BftBaseTest):
 
     def check_log(self, entity_or_radio: Union[env.ALEntity, env.Radio], regex: str,
-                  start_line: int = 0, timeout: float = 0.3) -> bool:
+                  start_line: int = 0, timeout: float = 0.6) -> bool:
```
Again, this should be in a separate commit.

arnout commented on tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_prplwrt.py:
```diff
@@ -23,8 +25,8 @@
-    model = ("prplWRT")
-    prompt = ['root\\@OpenWrt:/#', '/#', '@OpenWrt:/#']
+    model = "prplWRT"
+    prompt = ['root\\@OpenWrt:/#', '/#', '@OpenWrt:/#', "@OpenWrt:~#"]
```
I usually say that commits should be split, but this time it should be squashed with the first commit that updates prplmesh_prplwrt.py.

arnout commented on tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_prplwrt.py:
```diff
@@ -83,10 +115,12 @@ def check_status(self) -> bool:
-                ["operational", "FAIL", pexpect.EOF, pexpect.TIMEOUT],
-                timeout=10)
-        if match == 0:
+        self.expect(
+                ["(?P<main_agent>OK) Main agent.+"
+                 "(?P<wlan0>OK) wlan0.+"
+                 "(?P<wlan2>OK) wlan2", pexpect.TIMEOUT],
+                timeout=5)
```
You can actually just do
```
self.sendline("/etc/init.d/prplmesh status && echo "FULLY OPERATIONAL")
self.expect(["FULLY OPERATIONAL", pexpect.TIMEOUT], timeout=5)
```

abelog replied:
Just remembered, that i have seen that `/etc/init.d/prplmesh status` actually sets appropriate shell exit code. Since functionality is the same, readability same (at least to me) don't see much sense changing.

arnout reviewed (approved):
This PR is doing too many things, adding too many features, making the whole thing pretty hard to follow...

Given this situation, it's probably better to merge it more or less as-is, and fix things up later. I have a few comments already that you can take into account, but most likely I'll have more architectural comments in future PRs.

In the future, please make the changes in a single PR much smaller and easier to parse.

arnout commented:
> Approving because I think that with these big PRs, it's generally faster to just get them merged and fix whatever is still wrong or missing in a followup PR.

I agree with that, but the reviews should still happen and if it's already merged, *nobody* is going to review it...

Or did you do a full review and had no comments other than what I mentioned already?
